### DISCUSSION TITLE
Positive watts but negative amps?

### DISCUSSION THREAD

**Discussion by SteveMarquess2** [David00/rpi-power-monitor#134] [Q&A]
New user here. trying out what will be a fairly elaborate setup for a topologically complex electric service (muliple buildings, two PV ground mount systems, hybrid inverter, two generators, etc.)

Initial setup is one RPi power monitor with four CTs monitoring two 200A service entrances (in parallel downstream of 320/400A utility service).  CTs 1 and 2 are the phases for one service entrance, CTs 3 and 4 for the other. Both SEs are backfed by various inverters and so should run negative in the daytime. In these "Last 12 hours" shots notice all four CTs show positive power in the wee hours:

![12hrs-w](https://github.com/user-attachments/assets/9b994c00-8e74-44da-91cb-8d8ee15fb7f3)

So why does CT4 show negative amps?

![12hrs-a](https://github.com/user-attachments/assets/e641ccda-b5f1-4114-970e-66b14268e93b)

There's probably something simple I'm missing (after dodging it till now COVID finally got me, I've gone from digging trenches for conduit last week to barely able to walk). Any insights appreciated.

**Comment by David00** [comment]
Hi Steve, thanks for your patience - I was out of town but I'm back now!  I suspect this is due to the use of `reversed = true` in the config, and I might need to revisit the code about how `reversed = true` is handled. Can you share your config.toml file?

**Reply by SteveMarquess2** [reply]
Yep, definitely a problem with the `reversed = true` logic: after my first post I studied power_monitor.py a bit and couldn't make sense of that section of code. So to simplify analysis I physically flipped the CTs (three, IIRC) that had  `reversed = true`  and set  `reversed = false` for all CTs. And voila, suddenly everything (amps and watts) is showing up positive.

Incidentally when initially installing the CTs I was careful to have the raised arrows all pointing to the line side, which didn't seem to have much bearing on the polarity.

**Reply by David00** [reply]
Yep! I think I completely spaced one of the potential cases when handling `reversed = true`. It gets a little tricky when determining the net power of the various ct "types" (`mains`, `production`, `consumption`).  I've written a fix for this and just pushed it, so you can get it with git like so:

```bash
cd ~/rpi_power_monitor
git fetch
git pull
```

Since your installation looks to be a bit tricky, here is some additional context about the CT types:

- `mains` : intended for use on the service entrance / panel main conductors. Positive values mean power is flowing from the grid to the downstream loads. Negative values mean there is excess production downstream and power is flowing back to the grid.

- `production` : intended for use only on circuits that produce power. Should read positive when producing.

- `consumption`: intended for all other circuits that are _not_ mains or production... so the typical circuits for appliances, lighting, rooms, etc.

In your installation with four CTs on the service entrances, you'll only be able to see the net power.  To get the total consumption and production, you'll have to monitor all sources of production from downstream with a dedicated CT (you can use one CT for a two-pole circuit by setting `two_pole = true` in the config).

If your production sources do not all converge at a single panel, you may want to look into setting up a [remote Influx database](https://david00.github.io/rpi-power-monitor/docs/v0.3.0/advanced-usage.html#using-a-remote-influxdb-server) so you can dump the data from multiple power monitors into it (and potentially do some post-processing on it through the use of Influx continuous queries).

**Reply by SteveMarquess2** [reply]
I pulled the new power_monitor.py, thanks. Though I'm not going to switch the CTs back to test it.

I have already set up a remote Influx DB though haven't tied into it yet. My current four CTs are all "type = 'mains'" which is working as expected. I have both load side PV interconnects and supply side (upstream of the service disconnects) PV interconnects. Each of those sets of inverters will get its own rpi power monitor (for a total of three), and all those CTs will be "type = 'production'" (I won't have any 'consumption' monitoring, at least initially; I already have 14 traditional KWH meters around the premises for that).

My end goal is to be able to see and track a) net power consumption at the service point, b) total PV production, and c) to calculate premises power requirements exclusive of the PV production. There are a lot of grid outages here and I have a "almost" whole house automatic backup (hybrid inverter, battery bank, autostart diesel generator, custom active load management). The battery bank is due for replacement (I'll do something LFP) and having recently relocated the service point I'm considering going for true whole house backup.

Yup, a bit tricky. I've tried and discarded a couple of commercial monitoring gizmos. Your product is appealing because it seems to have the flexibility to configure the end result I want, at the cost of getting in the weeds a bit. Also I can use hardwired Ethernet exclusively (I'm using PoE hats) and avoiding wifi hell.

**Reply by David00** [reply]
Sounds good, Steve! I'm glad that this looks like it's going to work for your needs.
